### PR TITLE
Implement `.add_predictor`

### PR DESCRIPTION
**Pull request by martbern**

### CODE REVIEW

**Comment by martbern** [comment]
This closes #26 as well.

**Comment by martbern** [comment]
@bokajgd Jeg ved det er lidt op ad bakke at blive smidt ind i et code-review nu, men tænker det er en god måde at lære code-basen at kende på. Giv endelig lyd, hvis noget er uforståeligt.

File: src/timeseriesflattener/flattened_dataset.py [hunk 1]
```diff
@@ -27,69 +27,105 @@ def add_outcome(
...
             outcome_df (DataFrame): Cols: dw_ek_borger, datotid, value if relevant.
             lookahead_days (float): How far ahead to look for an outcome in days. If none found, use fallback.
-            resolve_multiple (str): What to do with more than one value within the lookahead.
-                Suggestions: earliest, latest, mean, max, min.
-            fallback (List[str]): What to do if no value within the lookahead.
-                Suggestions: latest, mean_of_patient, mean_of_population, hardcode (qualified guess)
-            values_colname (str): Colname for outcome values in outcome_df
+            resolve_multiple (Callable): How to handle multiple values within the lookahead window. Takes a a function that takes a list as an argument and returns a float.
+            fallback (float): What to do if no value within the lookahead.
+            source_values_colname (str): Colname for outcome values in outcome_df. Defaults to "val".
```
Comment (KennethEnevoldsen) [inline src/timeseriesflattener/flattened_dataset.py:42]: Not entirely sure what this argument values. Is it the column containing values? (assuming there is only one column that does this in a long format style?) or is it the column of outcomes?

File: src/timeseriesflattener/flattened_dataset.py [hunk 2]
```diff
@@ -27,69 +27,105 @@ def add_outcome(
...
-            ),
-            axis=1,
+        self._add_col(
+            values_df=outcome_df,
+            direction="ahead",
+            interval_days=lookahead_days,
+            resolve_multiple=resolve_multiple,
+            fallback=fallback,
+            new_col_name=new_col_name,
+            source_values_colname=source_values_colname,
```
Comment (KennethEnevoldsen) [inline src/timeseriesflattener/flattened_dataset.py:53]: I have no idea what this does - it adds a new column - a comment could probably do.

File: src/timeseriesflattener/flattened_dataset.py [hunk 3]
```diff
@@ -27,69 +27,105 @@ def add_outcome(
         self,
         outcome_df: DataFrame,
         lookahead_days: float,
-        resolve_multiple: str,
-        fallback: List[str],
-        values_colname: str = "values",
+        resolve_multiple: Callable,
+        fallback: float,
+        source_values_colname: str = "val",
```
Comment (KennethEnevoldsen) [inline src/timeseriesflattener/flattened_dataset.py:32]: Maybe a better argument name?

File: src/timeseriesflattener/flattened_dataset.py [hunk 4]
```diff
@@ -27,69 +27,105 @@ def add_outcome(
...
         predictor_df: DataFrame,
         lookbehind_days: float,
         resolve_multiple: str,
-        fallback: List[str],
-        outcome_colname: str,
+        fallback: float,
+        source_values_colname: str = "val",
+        new_col_name: str = None,
     ):
         """Adds a predictor-column to the dataset
```
Comment (KennethEnevoldsen) [inline src/timeseriesflattener/flattened_dataset.py:65]: What is a predictor-column?

File: src/timeseriesflattener/flattened_dataset.py [hunk 5]
```diff
@@ -27,69 +27,105 @@ def add_outcome(
...
+
+        Args:
+            values_df (DataFrame): Cols: dw_ek_borger, datotid, value.
+            direction (str): Whether to look "ahead" or "behind".
+            interval_days (float): How far to look in direction.
+            resolve_multiple (Callable): How to handle multiple values within the lookbehind window. Takes a a function that takes a list as an argument and returns a float.
             fallback (List[str]): What to do if no value within the lookahead.
-                Suggestions: latest, mean_of_patient, mean_of_population, hardcode (qualified guess)
-            outcome_colname (str): What to name the column
+            new_col_name (str): Name to use for new col. Automatically generated as '{new_col_name}_within_{lookahead_days}_days'.
```
Comment (KennethEnevoldsen) [inline src/timeseriesflattener/flattened_dataset.py:103]: `col_name` or `colname`?

Comment (KennethEnevoldsen) [reply src/timeseriesflattener/flattened_dataset.py:103]: In the description I think it is fair to write out column ;)

File: src/timeseriesflattener/flattened_dataset.py [hunk 6]
```diff
@@ -27,69 +27,105 @@ def add_outcome(
...
+        self,
+        values_df: DataFrame,
+        direction: str,
+        interval_days: float,
+        resolve_multiple: str,
+        fallback: float,
+        new_col_name: str,
+        source_values_colname: str = "val",
+    ):
+        """Adds a value-column to the dataset
```
Comment (KennethEnevoldsen) [inline src/timeseriesflattener/flattened_dataset.py:95]: What is value-column?

**Review by KennethEnevoldsen — COMMENTED** [review]
It seems like there is a lot of internal lingo, which is a bit hard to understand coming to the codebase. I think I would try to make those things more clear.

Another thing which is probably nice to take here is describe what a "flattened" time series is.

File: src/timeseriesflattener/flattened_dataset.py (hunk 3 above)
Comment (martbern) [reply src/timeseriesflattener/flattened_dataset.py:32]: Good point, attempted one :-)

File: src/timeseriesflattener/flattened_dataset.py (hunk 1 above)
Comment (martbern) [reply src/timeseriesflattener/flattened_dataset.py:42]: "Long format style"? 

Currently, the values must be floats or ints – any handling of e.g. strings has to happen before flattening. 

I've expanded the docstring.

File: src/timeseriesflattener/flattened_dataset.py (hunk 2 above)
Comment (martbern) [reply src/timeseriesflattener/flattened_dataset.py:53]: Renamed the method. Since `add_predictor` and `add_outcome` share such a large amount of code, I decided to have an internal method with a direction argument.

I still think having to user-facing methods is a good idea, to make the code easier to interpret.

File: src/timeseriesflattener/flattened_dataset.py (hunk 4 above)
Comment (martbern) [reply src/timeseriesflattener/flattened_dataset.py:65]: Expanded docstring

File: src/timeseriesflattener/flattened_dataset.py (hunk 6 above)
Comment (martbern) [reply src/timeseriesflattener/flattened_dataset.py:95]: Expanded the docstring

File: src/timeseriesflattener/flattened_dataset.py (hunk 5 above)
Comment (martbern) [reply src/timeseriesflattener/flattened_dataset.py:103]: Haha, fair! I've improved consistency.

File: src/timeseriesflattener/flattened_dataset.py (hunk 1 above)
Comment (KennethEnevoldsen) [reply src/timeseriesflattener/flattened_dataset.py]: [long format and wide format](https://www.statology.org/long-vs-wide-data/) are two ways to represent data. Long format is special because it forces only one column with values.

File: src/timeseriesflattener/flattened_dataset.py (hunk 2 above)
Comment (KennethEnevoldsen) [reply src/timeseriesflattener/flattened_dataset.py]: I do to, but the `_add_col` could basically be any kind of function even if it is internal it should be possible to interpret what the function does.

File: src/timeseriesflattener/flattened_dataset.py [hunk 7]
```diff
@@ -7,101 +7,157 @@ class FlattenedDataset:
...
+        And a time-series of blood-pressure values as an outcome:
+        id_col_name | timestamp_col_name | blood_pressure_value
+        1           | 2022-01-09         | 120
+        1           | 2022-01-14         | 140
+
+        Then you can "flatten" the outcome into a new df, with a row for each of your prediction times:
+        id_col_name | timestamp_col_name | latest_blood_pressure_within_24h
+        1           | 2022-01-10         | 120
+        1           | 2022-01-12         | NA
+        1           | 2022-01-15         | 140
```
Comment (KennethEnevoldsen) [inline src/timeseriesflattener/flattened_dataset.py:32]: So it is like a time-dependent merge.

Comment (KennethEnevoldsen) [reply src/timeseriesflattener/flattened_dataset.py:32]: `lookahead_merge()` and `lookbehind_merge()` (lookbehind is better than lookback as  seems to better explain to be what the function intends to do.

And then the internal function can be a `_lookaround_merge`

Comment (martbern) [reply src/timeseriesflattener/flattened_dataset.py:32]: Yeah, that's probably better! Added an issue for it.

Comment (KennethEnevoldsen) [reply src/timeseriesflattener/flattened_dataset.py:32]: I btw don't think the internal function `_lookaround_merge` what now called `_add_col` should not be internal. I could imagine that e.g. I would like to call that directly.

Comment (martbern) [reply src/timeseriesflattener/flattened_dataset.py:32]: That sentence is pretty hard to parse ;-)

I take it you think that `lookaround_merge` (which is currently called `add_col`) shouldn't be internal? Sure, I'll change that in main as well.

Comment (KennethEnevoldsen) [reply src/timeseriesflattener/flattened_dataset.py:32]: Yep that is what I meant (y)
